src/ForeignKeyConstraint.php

```diff
@@ -18,21 +18,64 @@ class ForeignKeyConstraint extends Constraint
     /**
      * @var string|null referenced table schema name.
      */
-    public $foreignSchemaName;
+    private $foreignSchemaName;
     /**
      * @var string referenced table name.
      */
-    public $foreignTableName;
+    private $foreignTableName;
     /**
      * @var string[] list of referenced table column names.
      */
-    public $foreignColumnNames;
+    private $foreignColumnNames;
     /**
      * @var string|null referential action if rows in a referenced table are to be updated.
      */
-    public $onUpdate;
+    private $onUpdate;
     /**
      * @var string|null referential action if rows in a referenced table are to be deleted.
      */
-    public $onDelete;
+    private $onDelete;
+
+    /**
+     * Constructor
+     * @param string|null name (inherited from parent) the constraint name.
+     * @param string[]|null columnNames (inherited from parent) list of column names the constraint belongs to.
+     * @param string|null foreignSchemaName referenced table schema name.
+     * @param string foreignTableName referenced table name.
+     * @param string[] foreignColumnNames list of referenced table column names.
+     * @param string|null onUpdate referential action if rows in a referenced table are to be updated.
+     * @param string|null onDelete referential action if rows in a referenced table are to be deleted.
+     */
+    public function __construct(?string $name, ?iterable $columnNames, ?string $foreignSchemaName, string $foreignTableName, iterable $foreignColumnNames, ?string $onDelete, ?string $onUpdate)
```

Comment:
It looks like big regression to me. Previously you have very readable key-based configuration array, now you need to remember order of 7 constructor arguments to understand what is what.

Reply:
I have thought these two points:
- avoid different syntax and approach with other *Constraint classes;
- how many times you use ForeignKeyConstraint directly to have to pass every parameters?

If direction is to have constructor based classes, I think that in this case we could go on.

Reply:
Why all these are null-able?

Reply:
Optional parameters may go into setters.

Reply:
If I put null-able parameters out of construtor parameters, the syntax to create and initialize object can be even harder, because developer should be:

1) Create the object with : `new ForeignKeyConstraint`, passing only needed parameters;
2) Set eventually optional parameters, one by one; here we have 4 null-able parameters, it  will need to write, eventualy, other 4 assigments;

Reply:
And what about Constraint construtor? It has all 2 fields null-able. 

Do we want to make a constructor without parameters or anyway we use parameters in constructor?

This could be a guideline. We could create the constructor for only not null-able parameters, and others filled using setters.

What do you thing?

Reply:
Thought about it more. Since the class isn't used by end users, the most straightforward way is to keep old array-based constructor.

Reply:
This class is a good example where I would prefer to extend from `ConfigurableObject`.
https://github.com/yiisoft/yii-core/pull/136